File: csharp-fp3/Domain/Account.cs
using CsharpFp3.Library;

namespace CsharpFp3.Domain;

public sealed record Account(Guid Id, Money Balance);

public abstract record AccountError
{
    public sealed record OpeningBalanceMustBeNonNegative : AccountError;

    public sealed record AmountMustBePositive : AccountError;

    public sealed record InsufficientBalance(Money Balance, Money Amount) : AccountError;

    public sealed record AccountNotFound(Guid AccountId) : AccountError;
}

public static class AccountDomain
{
    public static Result<Account, AccountError> Open(Guid id, Money openingBalance) =>
        openingBalance.Amount < 0m
            ? Result<Account, AccountError>.Fail(new AccountError.OpeningBalanceMustBeNonNegative())
            : Result<Account, AccountError>.Ok(new Account(id, openingBalance));

    public static Result<Account, AccountError> Withdraw(Account account, Money amount) =>
        (account.Balance.Amount, amount.Amount) switch
        {
            (_, <= 0m) => Result<Account, AccountError>.Fail(
                new AccountError.AmountMustBePositive()
            ),

            var (balance, transaction) when balance < transaction => Result<
                Account,
                AccountError
            >.Fail(new AccountError.InsufficientBalance(account.Balance, amount)),

            var (balance, transaction) => Result<Account, AccountError>.Ok(
                account with
                {
                    Balance = new Money(balance - transaction),
                }
            ),
        };
}

File: js-fp3/domain/account.js
import { ok, fail } from '../library/result.js';
import { match, when, any } from '../library/patterns.js';

export const openingBalanceMustBeNonNegative = { type: 'OpeningBalanceMustBeNonNegative' };

export const amountMustBePositive = { type: 'AmountMustBePositive' };

export const insufficientBalance = (balance, amount) => ({
  type: 'InsufficientBalance',
  balance,
  amount,
});

export const accountNotFound = (accountId) => ({
  type: 'AccountNotFound',
  accountId,
});

export const openAccount = (id, openingBalance) =>
  openingBalance.amount < 0
    ? fail(openingBalanceMustBeNonNegative)
    : ok({ id, balance: openingBalance });

export const withdraw = (account, amount) =>
  match(
    when(({ amount: a }) => a.amount <= 0, () => fail(amountMustBePositive)),
    when(
      ({ account: acc, amount: amt }) => acc.balance.amount < amt.amount,
      ({ account: acc, amount: amt }) => fail(insufficientBalance(acc.balance, amt))
    ),
    when(any, ({ account: acc, amount: amt }) =>
      ok({
        ...acc,
        balance: { amount: acc.balance.amount - amt.amount },
      })
    )
  )({ account, amount });
